File: crates/lunarbase-math/src/quote.rs
//! High-level direct and routed quote evaluation.

use crate::arithmetic::{checked_add, checked_sub};
use crate::fees::{
    calculate_fee_bps_for_router, quote_lane_exact_in, quote_lane_exact_in_fee,
    quote_lane_exact_out, quote_lane_exact_out_fee, quote_lane_slippage_bps,
    quote_lane_weighted_slippage_k_bps, split_fee,
};
use crate::slot0::{
    lane_slot0_ask_fee_bps, lane_slot0_bid_fee_bps, lane_slot0_block_delay,
    lane_slot0_latest_update_block, lane_slot0_price, lane_slot0_slippage_k_bps,
};
use crate::state::{
    LaneState, QuoteError, QuoteMode, QuoteOutcome, QuoteRequest, QuoteResult, QuoteState,
    UnavailableReason,
};
use crate::types::{Address, MathError, U256};

fn lane_or_reason(
    state: &QuoteState,
    asset: Address,
    execution_block_number: u64,
) -> Result<&LaneState, UnavailableReason> {
    let lane = state
        .lanes
        .get(&asset)
        .ok_or(UnavailableReason::MissingLane(asset))?;
    if !lane.exists() {
        return Err(UnavailableReason::MissingLane(asset));
    }
    if lane.paused() {
        return Err(UnavailableReason::PausedLane(asset));
    }
    let expires_at = lane_slot0_latest_update_block(lane.slot0)
        .saturating_add(u64::from(lane_slot0_block_delay(lane.slot0)));
    if execution_block_number > expires_at {
        return Err(UnavailableReason::StaleLane(asset));
    }
    Ok(lane)
}
fn principal_cash_value(lane: &LaneState) -> Result<U256, MathError> {
    let principal = U256::from(lane.total_principal_amount);
    if principal == U256::ZERO {
        return Ok(U256::ZERO);
    }
    quote_lane_exact_in(lane_slot0_price(lane.slot0), principal, false)
}
fn lane_spread(
    anchor: U256,
    fee_bps: U256,
    slippage_bps: U256,
    exact_in: bool,
) -> Result<(U256, U256), MathError> {
    let fee = if exact_in {
        quote_lane_exact_in_fee(anchor, fee_bps)?
    } else {
        quote_lane_exact_out_fee(anchor, fee_bps)?
    };
    let total_bps = checked_add(fee_bps, slippage_bps)?;
    let total = if exact_in {
        quote_lane_exact_in_fee(anchor, total_bps)?
    } else {
        quote_lane_exact_out_fee(anchor, total_bps)?
    };
    Ok((fee, checked_sub(total, fee)?))
}
fn assemble_quote(
    state: &QuoteState,
    request: &QuoteRequest,
    fee_asset: Address,
    anchor: U256,
    fee: U256,
    slippage_amount: U256,
) -> Result<QuoteOutcome, MathError> {
    let total_spread = checked_add(fee, slippage_amount)?;
    let (amount_in, amount_out) = match request.mode {
        QuoteMode::ExactIn => {
            if total_spread >= anchor {
                return Ok(QuoteOutcome::Unavailable(
                    UnavailableReason::SpreadConsumesAnchor,
                ));
            }
            (request.amount, checked_sub(anchor, total_spread)?)
        }
        QuoteMode::ExactOut => (checked_add(anchor, total_spread)?, request.amount),
    };
    let output_reserve = if request.asset_out == state.cash {
        state.cash_reserve
    } else {
        match state.lanes.get(&request.asset_out) {
            Some(lane) => lane.asset_reserve,
            None => {
                return Ok(QuoteOutcome::Unavailable(UnavailableReason::MissingLane(
                    request.asset_out,
                )));
            }
        }
    };
    let output_reserve = U256::from(output_reserve);
    let insufficient = amount_out > output_reserve
        || (request.mode == QuoteMode::ExactIn && fee > checked_sub(output_reserve, amount_out)?);
    if insufficient {
        return Ok(QuoteOutcome::Unavailable(
            UnavailableReason::InsufficientOutputReserve(request.asset_out),
        ));
    }
    let (partner, treasury) = split_fee(fee, state.fee_profile.partner_fee_bps(fee_asset))?;
    Ok(QuoteOutcome::Available(QuoteResult {
        amount_in,
        amount_out,
        fee_asset,
        fee_amount: fee,
        partner_fee: partner,
        treasury_fee: treasury,
    }))
}
fn quote_direct(
    state: &QuoteState,
    request: &QuoteRequest,
    lane_asset: Address,
    lane: &LaneState,
    cash_to_asset: bool,
    fee_asset: Address,
) -> Result<QuoteOutcome, MathError> {
    let anchor = match request.mode {
        QuoteMode::ExactIn => {
            quote_lane_exact_in(lane_slot0_price(lane.slot0), request.amount, cash_to_asset)?
        }
        QuoteMode::ExactOut => {
            quote_lane_exact_out(lane_slot0_price(lane.slot0), request.amount, cash_to_asset)?
        }
    };
    if anchor == U256::ZERO {
        return Ok(QuoteOutcome::Unavailable(UnavailableReason::ZeroAnchor));
    }
    let principal = principal_cash_value(lane)?;
    if principal == U256::ZERO {
        return Ok(QuoteOutcome::Unavailable(UnavailableReason::ZeroPrincipal(
            lane_asset,
        )));
    }
    let raw_fee = if cash_to_asset {
        lane_slot0_ask_fee_bps(lane.slot0)
    } else {
        lane_slot0_bid_fee_bps(lane.slot0)
    };
    let fee_bps = calculate_fee_bps_for_router(
        state.fee_profile.whitelisted,
        state.fee_profile.blacklist_fee_multiplier,
        raw_fee,
    )?;
    let swap_cash = if cash_to_asset {
        if request.mode == QuoteMode::ExactIn {
            request.amount
        } else {
            anchor
        }
    } else if request.mode == QuoteMode::ExactIn {
        anchor
    } else {
        request.amount
    };
    let slippage = quote_lane_slippage_bps(
        swap_cash,
        principal,
        U256::from(lane_slot0_slippage_k_bps(lane.slot0)),
    )?;
    let (fee, slippage_amount) = lane_spread(
        anchor,
        fee_bps,
        slippage,
        request.mode == QuoteMode::ExactIn,
    )?;
    assemble_quote(state, request, fee_asset, anchor, fee, slippage_amount)
}
fn quote_route(
    state: &QuoteState,
    request: &QuoteRequest,
    input_lane: &LaneState,
    output_lane: &LaneState,
) -> Result<QuoteOutcome, MathError> {
    let (intermediate_cash, anchor) = match request.mode {
        QuoteMode::ExactIn => {
            let intermediate =
                quote_lane_exact_in(lane_slot0_price(input_lane.slot0), request.amount, false)?;
            (
                intermediate,
                quote_lane_exact_in(lane_slot0_price(output_lane.slot0), intermediate, true)?,
            )
        }
        QuoteMode::ExactOut => {
            let intermediate =
                quote_lane_exact_out(lane_slot0_price(output_lane.slot0), request.amount, true)?;
            (
                intermediate,
                quote_lane_exact_out(lane_slot0_price(input_lane.slot0), intermediate, false)?,
            )
        }
    };
    if anchor == U256::ZERO {
        return Ok(QuoteOutcome::Unavailable(UnavailableReason::ZeroAnchor));
    }
    let first_principal = principal_cash_value(input_lane)?;
    if first_principal == U256::ZERO {
        return Ok(QuoteOutcome::Unavailable(UnavailableReason::ZeroPrincipal(
            request.asset_in,
        )));
    }
    let second_principal = principal_cash_value(output_lane)?;
    if second_principal == U256::ZERO {
        return Ok(QuoteOutcome::Unavailable(UnavailableReason::ZeroPrincipal(
            request.asset_out,
        )));
    }
    let weighted_k = quote_lane_weighted_slippage_k_bps(
        first_principal,
        U256::from(lane_slot0_slippage_k_bps(input_lane.slot0)),
        second_principal,
        U256::from(lane_slot0_slippage_k_bps(output_lane.slot0)),
    )?;
    let slippage = quote_lane_slippage_bps(
        intermediate_cash,
        checked_add(first_principal, second_principal)?,
        weighted_k,
    )?;
    let whitelisted = state.fee_profile.whitelisted;
    let bid = calculate_fee_bps_for_router(
        whitelisted,
        state.fee_profile.blacklist_fee_multiplier,
        lane_slot0_bid_fee_bps(input_lane.slot0),
    )?;
    let ask = calculate_fee_bps_for_router(
        whitelisted,
        state.fee_profile.blacklist_fee_multiplier,
        lane_slot0_ask_fee_bps(output_lane.slot0),
    )?;
    let fee_bps = checked_add(bid, ask)?;
    let fee = if request.mode == QuoteMode::ExactIn {
        quote_lane_exact_in_fee(anchor, fee_bps)?
    } else {
        quote_lane_exact_out_fee(anchor, fee_bps)?
    };
    let total_bps = checked_add(checked_add(bid, slippage)?, checked_add(ask, slippage)?)?;
    let total = if request.mode == QuoteMode::ExactIn {
        quote_lane_exact_in_fee(anchor, total_bps)?
    } else {
        quote_lane_exact_out_fee(anchor, total_bps)?
    };
    let fee_asset = if request.mode == QuoteMode::ExactIn {
        request.asset_out
    } else {
        request.asset_in
    };
    assemble_quote(
        state,
        request,
        fee_asset,
        anchor,
        fee,
        checked_sub(total, fee)?,
    )
}
/// Calculates a complete quote for a direct lane or a two-leg CASH route.
///
/// The function applies the contract's zero/equal-asset sentinels, lane
/// validity predicate, anchor conversion, configured fee profile,
/// principal-based slippage, spread, output-reserve availability, and
/// partner/treasury split.
///
/// `execution_block_number` must be the EVM-visible block number supplied by
/// the runtime's normalized head, not an arbitrary provider sequence.
///
/// # Errors
///
/// Returns [`QuoteError::Arithmetic`] when Solidity's checked arithmetic
/// boundary is exceeded.
pub fn quote(
    request: &QuoteRequest,
    execution_block_number: u64,
    state: &QuoteState,
) -> Result<QuoteOutcome, QuoteError> {
    if request.amount == U256::ZERO {
        return Ok(QuoteOutcome::Unavailable(UnavailableReason::ZeroAmount));
    }
    if request.asset_in == request.asset_out {
        return Ok(QuoteOutcome::Unavailable(UnavailableReason::EqualAssets));
    }
    let fee_asset = if request.mode == QuoteMode::ExactIn {
        request.asset_out
    } else {
        request.asset_in
    };
    if request.asset_out == state.cash {
        let lane = lane_or_reason(state, request.asset_in, execution_block_number);
        return match lane {
            Ok(lane) => Ok(quote_direct(
                state,
                request,
                request.asset_in,
                lane,
                false,
                fee_asset,
            )?),
            Err(reason) => Ok(QuoteOutcome::Unavailable(reason)),
        };
    }
    if request.asset_in == state.cash {
        let lane = lane_or_reason(state, request.asset_out, execution_block_number);
        return match lane {
            Ok(lane) => Ok(quote_direct(
                state,
                request,
                request.asset_out,
                lane,
                true,
                fee_asset,
            )?),
            Err(reason) => Ok(QuoteOutcome::Unavailable(reason)),
        };
    }
    let input_lane = match lane_or_reason(state, request.asset_in, execution_block_number) {
        Ok(lane) => lane,
        Err(reason) => return Ok(QuoteOutcome::Unavailable(reason)),
    };
    let output_lane = match lane_or_reason(state, request.asset_out, execution_block_number) {
        Ok(lane) => lane,
        Err(reason) => return Ok(QuoteOutcome::Unavailable(reason)),
    };
    Ok(quote_route(state, request, input_lane, output_lane)?)
}
/// Converts a rich quote outcome to `Lanes.quoteExactIn`'s public scalar.
///
/// Available quotes return `amount_out`; every unavailable reason maps to zero.
pub fn solidity_exact_in_amount(outcome: &QuoteOutcome) -> U256 {
    match outcome {
        QuoteOutcome::Available(result) => result.amount_out,
        QuoteOutcome::Unavailable(_) => U256::ZERO,
    }
}
/// Converts a rich quote outcome to `Lanes.quoteExactOut`'s public scalar.
///
/// Available quotes return `amount_in`; unavailable results map to the
/// contract's `U256::MAX` sentinel.
pub fn solidity_exact_out_amount(outcome: &QuoteOutcome) -> U256 {
    match outcome {
        QuoteOutcome::Available(result) => result.amount_in,
        QuoteOutcome::Unavailable(_) => U256::MAX,
    }
}
/// Applies the special zero-request override to the exact-out sentinel.
///
/// Solidity returns zero for a zero requested amount, even though a generic
/// unavailable exact-out outcome normally maps to `U256::MAX`.
pub fn solidity_exact_out_amount_for_request(
    request: &QuoteRequest,
    outcome: &QuoteOutcome,
) -> U256 {
    if request.amount == U256::ZERO {
        U256::ZERO
    } else {
        solidity_exact_out_amount(outcome)
    }
}

File: packages/math/src/quote.ts
import { assertU256, U256_MAX, type Address } from "./constants.js";
import { checkedAdd, checkedSub } from "./arithmetic.js";
import {
  laneSlot0AskFeeBps,
  laneSlot0BidFeeBps,
  laneSlot0BlockDelay,
  laneSlot0LatestUpdateBlock,
  laneSlot0Price,
  laneSlot0SlippageKBps,
} from "./slot0.js";
import {
  calculateFeeBpsForRouter,
  quoteLaneExactIn,
  quoteLaneExactInFee,
  quoteLaneExactOut,
  quoteLaneExactOutFee,
  quoteLaneSlippageBps,
  quoteLaneWeightedSlippageKBps,
  splitFee,
} from "./fees.js";
import {
  laneExists,
  lanePaused,
  type FeeAllocation,
  type LaneState,
  type QuoteOutcome,
  type QuotePolicy,
  type QuoteRequest,
  type QuoteState,
  type UnavailableReason,
} from "./types.js";

function laneOrReason(state: QuoteState, asset: Address, executionBlockNumber: bigint): LaneState | UnavailableReason {
  const lane = state.lanes.get(asset.toLowerCase() as Address);
  if (!lane || !laneExists(lane)) return { kind: "MissingLane", asset };
  if (lanePaused(lane)) return { kind: "PausedLane", asset };
  const expiresAt = checkedAdd(laneSlot0LatestUpdateBlock(lane.slot0), BigInt(laneSlot0BlockDelay(lane.slot0)));
  if (executionBlockNumber > expiresAt) return { kind: "StaleLane", asset };
  return lane;
}

function principalCashValue(lane: LaneState): bigint {
  return lane.totalPrincipalAmount === 0n
    ? 0n
    : quoteLaneExactIn(laneSlot0Price(lane.slot0), lane.totalPrincipalAmount, false);
}

function laneSpread(anchor: bigint, feeBps: bigint, slippageBps: bigint, exactIn: boolean): readonly [bigint, bigint] {
  const fee = exactIn ? quoteLaneExactInFee(anchor, feeBps) : quoteLaneExactOutFee(anchor, feeBps);
  const totalBps = checkedAdd(feeBps, slippageBps);
  const total = exactIn ? quoteLaneExactInFee(anchor, totalBps) : quoteLaneExactOutFee(anchor, totalBps);
  return [fee, checkedSub(total, fee)];
}

function assembleQuote(
  state: QuoteState,
  policy: QuotePolicy,
  request: QuoteRequest,
  feeAsset: Address,
  anchor: bigint,
  fee: bigint,
  slippageAmount: bigint,
): QuoteOutcome {
  const totalSpread = checkedAdd(fee, slippageAmount);
  let amountIn: bigint;
  let amountOut: bigint;
  if (request.mode === "ExactIn") {
    if (totalSpread >= anchor) return { kind: "Unavailable", reason: { kind: "SpreadConsumesAnchor" } };
    amountIn = request.amount;
    amountOut = checkedSub(anchor, totalSpread);
  } else {
    amountIn = checkedAdd(anchor, totalSpread);
    amountOut = request.amount;
  }
  const outputReserve =
    request.assetOut.toLowerCase() === state.cash.toLowerCase()
      ? state.cashReserve
      : state.lanes.get(request.assetOut.toLowerCase() as Address)?.assetReserve;
  if (outputReserve === undefined)
    return { kind: "Unavailable", reason: { kind: "MissingLane", asset: request.assetOut } };
  const insufficient = amountOut > outputReserve || (request.mode === "ExactIn" && fee > outputReserve - amountOut);
  if (insufficient)
    return { kind: "Unavailable", reason: { kind: "InsufficientOutputReserve", asset: request.assetOut } };
  let feeAllocation: FeeAllocation | undefined;
  if (policy.verifiedPartnerFeeBps !== undefined) {
    const [partnerFee, treasuryFee] = splitFee(fee, BigInt(policy.verifiedPartnerFeeBps));
    feeAllocation = { partnerFee, treasuryFee };
  }
  return {
    kind: "Available",
    result: { amountIn, amountOut, feeAsset, feeAmount: fee, feeAllocation },
  };
}

function directQuote(
  state: QuoteState,
  policy: QuotePolicy,
  request: QuoteRequest,
  laneAsset: Address,
  lane: LaneState,
  cashToAsset: boolean,
  feeAsset: Address,
): QuoteOutcome {
  const anchor =
    request.mode === "ExactIn"
      ? quoteLaneExactIn(laneSlot0Price(lane.slot0), request.amount, cashToAsset)
      : quoteLaneExactOut(laneSlot0Price(lane.slot0), request.amount, cashToAsset);
  if (anchor === 0n) return { kind: "Unavailable", reason: { kind: "ZeroAnchor" } };
  const principal = principalCashValue(lane);
  if (principal === 0n) return { kind: "Unavailable", reason: { kind: "ZeroPrincipal", asset: laneAsset } };
  const rawFee = cashToAsset ? laneSlot0AskFeeBps(lane.slot0) : laneSlot0BidFeeBps(lane.slot0);
  const feeBps = calculateFeeBpsForRouter(policy.feeClass === "Whitelisted", state.blacklistFeeMultiplier, rawFee);
  const swapCash = cashToAsset
    ? request.mode === "ExactIn"
      ? request.amount
      : anchor
    : request.mode === "ExactIn"
      ? anchor
      : request.amount;
  const slippage = quoteLaneSlippageBps(swapCash, principal, BigInt(laneSlot0SlippageKBps(lane.slot0)));
  const [fee, slippageAmount] = laneSpread(anchor, feeBps, slippage, request.mode === "ExactIn");
  return assembleQuote(state, policy, request, feeAsset, anchor, fee, slippageAmount);
}

function routeQuote(
  state: QuoteState,
  policy: QuotePolicy,
  request: QuoteRequest,
  inputLane: LaneState,
  outputLane: LaneState,
): QuoteOutcome {
  let intermediateCash: bigint;
  let anchor: bigint;
  if (request.mode === "ExactIn") {
    intermediateCash = quoteLaneExactIn(laneSlot0Price(inputLane.slot0), request.amount, false);
    anchor = quoteLaneExactIn(laneSlot0Price(outputLane.slot0), intermediateCash, true);
  } else {
    intermediateCash = quoteLaneExactOut(laneSlot0Price(outputLane.slot0), request.amount, true);
    anchor = quoteLaneExactOut(laneSlot0Price(inputLane.slot0), intermediateCash, false);
  }
  if (anchor === 0n) return { kind: "Unavailable", reason: { kind: "ZeroAnchor" } };
  const firstPrincipal = principalCashValue(inputLane);
  if (firstPrincipal === 0n) return { kind: "Unavailable", reason: { kind: "ZeroPrincipal", asset: request.assetIn } };
  const secondPrincipal = principalCashValue(outputLane);
  if (secondPrincipal === 0n)
    return { kind: "Unavailable", reason: { kind: "ZeroPrincipal", asset: request.assetOut } };
  const weightedK = quoteLaneWeightedSlippageKBps(
    firstPrincipal,
    BigInt(laneSlot0SlippageKBps(inputLane.slot0)),
    secondPrincipal,
    BigInt(laneSlot0SlippageKBps(outputLane.slot0)),
  );
  const slippage = quoteLaneSlippageBps(intermediateCash, checkedAdd(firstPrincipal, secondPrincipal), weightedK);
  const whitelisted = policy.feeClass === "Whitelisted";
  const multiplier = state.blacklistFeeMultiplier;
  const bid = calculateFeeBpsForRouter(whitelisted, multiplier, laneSlot0BidFeeBps(inputLane.slot0));
  const ask = calculateFeeBpsForRouter(whitelisted, multiplier, laneSlot0AskFeeBps(outputLane.slot0));
  const feeBps = checkedAdd(bid, ask);
  const fee = request.mode === "ExactIn" ? quoteLaneExactInFee(anchor, feeBps) : quoteLaneExactOutFee(anchor, feeBps);
  const totalBps = checkedAdd(checkedAdd(bid, slippage), checkedAdd(ask, slippage));
  const total =
    request.mode === "ExactIn" ? quoteLaneExactInFee(anchor, totalBps) : quoteLaneExactOutFee(anchor, totalBps);
  const feeAsset = request.mode === "ExactIn" ? request.assetOut : request.assetIn;
  return assembleQuote(state, policy, request, feeAsset, anchor, fee, checkedSub(total, fee));
}

/**
 * Produces a bit-exact quote from one immutable in-memory state snapshot.
 *
 * `executionBlockNumber` is the EVM-visible block tracked by the runtime.
 * Pure math never reads RPC, persistence, wall-clock freshness, or a router.
 */
export function quote(
  request: QuoteRequest,
  executionBlockNumber: bigint,
  state: QuoteState,
  policy: QuotePolicy,
): QuoteOutcome {
  assertU256(request.amount, "amount");
  assertU256(executionBlockNumber, "executionBlockNumber");
  if (request.amount === 0n) return { kind: "Unavailable", reason: { kind: "ZeroAmount" } };
  if (request.assetIn.toLowerCase() === request.assetOut.toLowerCase())
    return { kind: "Unavailable", reason: { kind: "EqualAssets" } };
  const feeAsset = request.mode === "ExactIn" ? request.assetOut : request.assetIn;
  if (request.assetOut.toLowerCase() === state.cash.toLowerCase()) {
    const lane = laneOrReason(state, request.assetIn, executionBlockNumber);
    return "kind" in lane
      ? { kind: "Unavailable", reason: lane }
      : directQuote(state, policy, request, request.assetIn, lane, false, feeAsset);
  }
  if (request.assetIn.toLowerCase() === state.cash.toLowerCase()) {
    const lane = laneOrReason(state, request.assetOut, executionBlockNumber);
    return "kind" in lane
      ? { kind: "Unavailable", reason: lane }
      : directQuote(state, policy, request, request.assetOut, lane, true, feeAsset);
  }
  const input = laneOrReason(state, request.assetIn, executionBlockNumber);
  if ("kind" in input) return { kind: "Unavailable", reason: input };
  const output = laneOrReason(state, request.assetOut, executionBlockNumber);
  if ("kind" in output) return { kind: "Unavailable", reason: output };
  return routeQuote(state, policy, request, input, output);
}

/**
 * Converts a rich outcome to the scalar returned by the matching Solidity
 * quote function, including exact-out sentinel behavior.
 */
export function solidityQuoteAmount(request: QuoteRequest, outcome: QuoteOutcome): bigint {
  if (request.mode === "ExactIn") return outcome.kind === "Available" ? outcome.result.amountOut : 0n;
  if (request.amount === 0n) return 0n;
  return outcome.kind === "Available" ? outcome.result.amountIn : U256_MAX;
}
